### PR TITLE
Add gateway deployment for Pipelines as Code

### PR DESCRIPTION
**Pull request by fgiloux**
This provides the manifest and a workflow for deploying a gateway.
There are however limitations and issues:
- ingress is broken with kcp 0.5
- there is no controller watching EventListeners, which means that the gateway is statically configured for Pipelines as Code but in its current state does not support Tekton Triggers
- PipelineRuns triggered by PaC are not visible in any kcp workspace

These aspects will be addressed in phase 2.

Signed-off-by: Frederic Giloux <fgiloux@redhat.com>

### CODE REVIEW

File: gitops/sre/.tekton/gateway-deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        taskSpec:
+          workspaces:
+            - name: source
+          steps:
+            - name: gateway-deployment
+              image: ghcr.io/openshift-pipelines/gateway-deployment:main
+              workingDir: $(workspaces.source.path)
+              env:
+                - name: DATA_DIR
+                  value: $(workspaces.source.path)/gitops/sre
```
Comment (bnallapeta) [inline gitops/sre/.tekton/gateway-deployment.yaml:50]: We have set this value as  $(workspaces.source.path) for the other two images. And in the script, it is indeed referenced like this. So, could you please change the value for the cluster-setup.yaml and kcp-registration.yaml too?
files=($(ls $DATA_DIR/credentials/kubeconfig/compute))

File: images/gateway/install.sh [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	KUBECONFIG=${kcp_kcfg} kubectl kcp workspace use ${KCP_WORKSPACE}
+
+    else
+       printf "creating workspace %s\n" "${KCP_WORKSPACE}"
+       KUBECONFIG=${kcp_kcfg} kubectl kcp workspace create "${KCP_WORKSPACE}" --enter
+    fi
+}
+
+install_gateway () {
+    KUBECONFIG=${kcp_kcfg}
```
Comment (bnallapeta) [inline images/gateway/install.sh:92]: Since you are using KUBECONFIG directly in line 94, this can be removed.

File: images/gateway/install.sh [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    CONFIG_DIR="${DATA_DIR}/environment/kcp/gateway"
+    KUBECONFIG=${kcp_kcfg} kubectl apply -k  "${CONFIG_DIR}"
+}
+
+main() {
+printf "Switching to organization %s\n" ${KCP_ORG}
```
Comment (bnallapeta) [inline images/gateway/install.sh:98]: I think this line is here by mistake?

File: images/gateway/install.sh [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    prechecks
+    kcp_kubeconfig
+    printf "Switching to organization %s\n" ${KCP_ORG}
```
Comment (bnallapeta) [inline images/gateway/install.sh:101]: Double quote KCP_ORG, KCP_WORKSPACE everywhere

File: images/kcp-registrar/register.sh [hunk 5]
```diff
@@ -39,7 +39,7 @@ prechecks () {
         exit 1
     fi
 
-    KCP_ORG=${KCP_WORKSPACE:-}
+    KCP_WORKSPACE=${KCP_WORKSPACE:-}
```
Comment (bnallapeta) [inline images/kcp-registrar/register.sh:42]: This file is old. Rebasing should remove this file.

File: images/gateway/install.sh [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    else
+       printf "creating workspace %s\n" "${KCP_WORKSPACE}"
+       KUBECONFIG=${kcp_kcfg} kubectl kcp workspace create "${KCP_WORKSPACE}" --enter
+    fi
+}
+
+install_gateway () {
+    KUBECONFIG=${kcp_kcfg}
+    CONFIG_DIR="${DATA_DIR}/environment/kcp/gateway"
+    KUBECONFIG=${kcp_kcfg} kubectl apply -k  "${CONFIG_DIR}"
```
Comment (Roming22) [inline images/gateway/install.sh:94]: double space after `-k`.

File: images/gateway/install.sh (hunk 4 above)
Comment (Roming22) [inline images/gateway/install.sh:101]: I'd move the print statement in the function itself, but that might be a personal preference.

File: gitops/sre/.tekton/gateway-deployment.yaml (hunk 1 above)
Comment (bnallapeta) [reply gitops/sre/.tekton/gateway-deployment.yaml:50]: We do not need the gitops/sre as folders 'credentials' and 'environment' are placed at the root level.
Please ignore my previous comment. And in this file, we need to remove '/gitops/sre'.

Comment (fgiloux) [reply gitops/sre/.tekton/gateway-deployment.yaml:50]: > We have set this value as $(workspaces.source.path) for the other two images. And in the script, it is indeed referenced like this. So, could you please change the value for the cluster-setup.yaml and kcp-registration.yaml too?

This PR is focused on the gateway. It is unfortunate that changes made by PR #86 broke other things.

> And in this file, we need to remove '/gitops/sre'.

I disagree. The location of the credentials directory is under gitops/sre/credentials, which is exactly what you get if you fork this repository. Other approaches of creating a gitops repository, which may not require a gitops/sre directory are acceptable but then the persons creating such a repository can adjust the environment variable to their custom structure.

The documentation, which should have come with PR #86  is unfortunately missing. That's the reason I created [issue #105 ](https://github.com/openshift-pipelines/pipelines-service/issues/105). It is unfortunate that it has not been worked on.

File: images/gateway/install.sh (hunk 2 above)
Comment (fgiloux) [reply images/gateway/install.sh]: true. Thanks.

File: images/gateway/install.sh (hunk 3 above)
Comment (fgiloux) [reply images/gateway/install.sh]: True. Thanks

File: images/gateway/install.sh (hunk 4 above)
Comment (fgiloux) [reply images/gateway/install.sh]: ok

**Comment by fgiloux** [comment]
@bnallapeta @Roming22 anything preventing the merge of this PR?

File: gitops/sre/.tekton/gateway-deployment.yaml (hunk 1 above)
Comment (bnallapeta) [reply gitops/sre/.tekton/gateway-deployment.yaml:50]: - Currently, both the cluster-setup and kcp-registrar PR have workingDir: $(workspaces.source.path). 
- And even in the register.sh script for kcp-registrar, we are using $DATA_DIR/credentials/kubeconfig where DATA_DIR will point to wherever credentials and environment are present.

https://github.com/openshift-pipelines/pipelines-service/blob/main/gitops/sre/.tekton/kcp-registration.yaml#L47
https://github.com/openshift-pipelines/pipelines-service/blob/main/gitops/sre/.tekton/pln-svc-registration.yaml#L47

File: docs/gateway.md [hunk 7]
```diff
@@ -70,3 +70,10 @@ here the backend is the service `httpecho` in the `trigger` namespace listening
 **_NOTE:_**  This is only needed for phase 1. This component will get removed when we move to phase 2 and have the event listeners provisioned through kcp.
 
 ---
+
```
Comment (bnallapeta) [inline docs/gateway.md:73]: We also need to update content under Configuration section.
eg: server s1 httpecho.trigger.svc.cluster.local:80 -> server el-pac el-pipelines-as-code-interceptor.openshift-pipelines.svc.cluster.local:8080

Comment (fgiloux) [reply docs/gateway.md:73]: done

File: gitops/sre/.tekton/gateway-deployment.yaml (hunk 1 above)
Comment (fgiloux) [reply gitops/sre/.tekton/gateway-deployment.yaml]: changed it for the sake of consistency with the other PipelineRuns

**Comment by fgiloux** [comment]
@bnallapeta I should have addressed your requets

**Review by bnallapeta — APPROVED** [review]
LGTM. Merging the PR.

**Pull request merged**